File: LeetCode/19RemoveNthNodeFromEndOfList/RemoveNthNodeFromEndOfList.cpp
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        // Single pass -> Time:O(n), Space:O(1)
        ListNode* travel = head;
        ListNode* remove = head;
        ListNode* remove_prev = head;
        int count = 0;
        while (travel != NULL) {
            travel = travel->next;
            count++;
            if (count > n) {
                remove_prev = remove;
                remove = remove->next;
            }
        }
        if (remove == remove_prev) {
            head = head->next;
        }
        else {
            remove_prev->next = remove->next;
        }
        return head;
    }
};

File: LeetCode/19RemoveNthNodeFromEndOfList/RemoveNthNodeFromEndOfList.py
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # Single pass -> Time:O(n), Space:O(1)
        travel = head
        remove = head
        remove_prev = head
        count = 0
        while travel != None:
            travel = travel.next
            count += 1
            if count > n:
                remove_prev = remove
                remove = remove.next
        if remove_prev == remove:
            head = head.next
        else:
            remove_prev.next = remove.next
        return head
